pkg/dockerfile/buildargs.go

```diff
@@ -51,8 +52,13 @@ func (b *BuildArgs) Clone() *BuildArgs {
 
 // ReplacementEnvs returns a list of filtered environment variables
 func (b *BuildArgs) ReplacementEnvs(envs []string) []string {
+	// Ensure that we operate on a new array and do not modify the underlying array
+	resultEnv := make([]string, len(envs))
+	copy(resultEnv, envs)
 	filtered := b.FilterAllowed(envs)
-	return append(envs, filtered...)
+	resultEnv = append(resultEnv, filtered...)
+	sort.Strings(resultEnv)
```

Comment:
In general I'm not totally convinced that it's safe to sort env vars. I understand why we'd want that for cache key generation, but `FOO=x FOO=b echo $FOO` has demonstrably different results if you sort the env vars. The output of those should produce different cached results as well.

Am I misunderstanding the problem that this solves?

Reply:
@imjasonh, @apollo13, instead of sorting the result list here which would affect all callers of the `ReplacementEnvs` function (for which there seems to be many), couldn't the sort be done in the `populateCompositeKey` function in `pkg/executor/build.go` just before [line 184](https://github.com/apollo13/kaniko/blob/1dfda2bb812c0d87ae4b2d1372b261f3ef5ff685/pkg/executor/build.go#L184) when the result is used next?

That way, we know that the sorting will only be applied in the use cases of a cache key.

Reply:
I had that before, but I think just sorting `filtered` in here should be fine.

Reply:
Pushed a new commit moving the sort into `populateCompositeKey`